File: backend/app/models/denominator_change.py
"""
Denominator change audit trail models.

Models for detecting and documenting when the CAM reconciliation denominator
shifts between periods — RSF re-measurement, tenant roster changes,
self-maintenance conversions, BOMA standard changes, etc.
"""

from datetime import datetime
from decimal import Decimal
from enum import Enum
from uuid import UUID

from pydantic import BaseModel, Field


class DenominatorChangeType(str, Enum):
    """Types of denominator changes that can occur between reconciliation periods."""

    RSF_REMEASUREMENT = "rsf_remeasurement"
    TENANT_ADDED = "tenant_added"
    TENANT_REMOVED = "tenant_removed"
    SELF_MAINTENANCE_START = "self_maintenance_start"
    SELF_MAINTENANCE_STOP = "self_maintenance_stop"
    EXCLUSION_CHANGE = "exclusion_change"
    BOMA_STANDARD_CHANGE = "boma_standard_change"
    SHARE_RECALCULATION = "share_recalculation"


class DenominatorChange(BaseModel):
    """A single detected change in the reconciliation denominator."""

    change_type: DenominatorChangeType = Field(
        description="Category of the denominator change"
    )
    description: str = Field(description="Human-readable description of the change")
    prior_value: str = Field(description="Value in the prior period")
    current_value: str = Field(description="Value in the current period")
    impact_description: str = Field(
        description="How this change affects tenant pro-rata shares"
    )


class TenantShareImpact(BaseModel):
    """Per-tenant impact of denominator changes on pro-rata share and recovery."""

    lease_id: UUID = Field(description="Lease affected by the change")
    tenant_name: str = Field(description="Tenant name for display")
    prior_pro_rata_share: Decimal = Field(
        description="Pro-rata share in prior period (0-1)"
    )
    current_pro_rata_share: Decimal = Field(
        description="Pro-rata share in current period (0-1)"
    )
    share_delta_pct_points: Decimal = Field(
        description="Change in share as percentage points"
    )
    prior_estimated_recovery: Decimal = Field(
        description="Estimated recovery amount in prior period"
    )
    current_estimated_recovery: Decimal = Field(
        description="Estimated recovery amount in current period"
    )
    recovery_delta: Decimal = Field(description="Dollar change in estimated recovery")
    contributing_changes: list[DenominatorChangeType] = Field(
        description="Which denominator changes contributed to this impact"
    )


class DenominatorChangeReport(BaseModel):
    """Complete report of denominator changes between two reconciliation periods."""

    property_id: UUID = Field(description="Property being analyzed")
    property_name: str = Field(description="Property name for display")
    prior_period: str = Field(
        description="Prior period range, e.g. '2023-01-01 to 2023-12-31'"
    )
    current_period: str = Field(description="Current period range")
    prior_total_rsf: Decimal = Field(description="Total rentable SF in prior period")
    current_total_rsf: Decimal = Field(
        description="Total rentable SF in current period"
    )
    rsf_delta: Decimal = Field(description="Change in total RSF")
    rsf_delta_percent: Decimal = Field(description="Percentage change in RSF")
    changes: list[DenominatorChange] = Field(
        description="All detected denominator changes"
    )
    tenant_impacts: list[TenantShareImpact] = Field(
        description="Per-tenant impact of the changes"
    )
    summary: str = Field(description="Executive summary of changes")
    generated_at: datetime = Field(description="When this report was generated")
    comparison_available: bool = Field(
        default=True,
        description=(
            "False when there is no finalized snapshot to compare against "
            "(a normal, expected state). The route returns HTTP 200 with an "
            "otherwise-empty report in that case."
        ),
    )
    missing_period: str | None = Field(
        default=None,
        description=(
            "Which period lacks a finalized snapshot when comparison_available "
            "is false: 'current' or 'prior'. None for a real comparison."
        ),
    )

File: frontend/src/features/reconciliation/types/denominator-change.ts
/**
 * TypeScript interfaces for denominator change analysis.
 *
 * Mirrors backend models in app/models/denominator_change.py.
 */

export type DenominatorChangeType =
  | 'rsf_remeasurement'
  | 'tenant_added'
  | 'tenant_removed'
  | 'self_maintenance_start'
  | 'self_maintenance_stop'
  | 'exclusion_change'
  | 'boma_standard_change'
  | 'share_recalculation'

export interface DenominatorChange {
  change_type: DenominatorChangeType
  description: string
  prior_value: string
  current_value: string
  impact_description: string
}

export interface TenantShareImpact {
  lease_id: string
  tenant_name: string
  prior_pro_rata_share: number
  current_pro_rata_share: number
  share_delta_pct_points: number
  prior_estimated_recovery: number
  current_estimated_recovery: number
  recovery_delta: number
  contributing_changes: DenominatorChangeType[]
}

export interface DenominatorChangeReport {
  property_id: string
  property_name: string
  prior_period: string
  current_period: string
  prior_total_rsf: number
  current_total_rsf: number
  rsf_delta: number
  rsf_delta_percent: number
  changes: DenominatorChange[]
  tenant_impacts: TenantShareImpact[]
  summary: string
  generated_at: string
  /**
   * False when there is no finalized snapshot to compare against (a normal,
   * expected state --- not a failure). The backend returns HTTP 200 with an
   * otherwise-empty report so the panel can guide the user instead of relying
   * on a 4xx. Defaults to true for a real comparison.
   */
  comparison_available: boolean
  /**
   * Which period is missing its finalized snapshot when
   * `comparison_available` is false: 'current' (this year not finalized) or
   * 'prior' (last year not finalized). Null/absent for a real comparison.
   */
  missing_period?: 'current' | 'prior' | null
}

export interface DenominatorChangeRequest {
  property_id: string
  current_period_start: string
  current_period_end: string
  prior_period_start?: string
  prior_period_end?: string
  prior_total_rsf?: number
  current_total_rsf?: number
}

export interface DenominatorChangePdfRequest {
  property_id: string
  current_period_start: string
  current_period_end: string
  prior_period_start?: string
  prior_period_end?: string
  prior_total_rsf?: number
  current_total_rsf?: number
}
